### PR TITLE
Fix TaskInstance actions with upstream/downstream

### PR DESCRIPTION
**Pull request by ashb**
When we added clearing individual mapped tasks, we unfortunately broke
the up/down stream feature

This was because when passing task_id/task_id+map_index down that
limited it to _just_ that task_id.

So we need to change it to also support the up/downstream we need to add
those tasks to the list we pass on, meaning that we have to support
task_id and (task_id,map_index) tuples in the same `task_id` list.

This almost certainly needs tests adding!

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragement file, named `{pr_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/models/dag.py [hunk 1]
```diff
@@ -1439,19 +1439,34 @@ def _get_task_instances(
...
             tis = tis.filter(TaskInstance.dag_id == self.dag_id, TaskInstance.task_id.in_(self.task_ids))
+        else:
+            tis = tis.filter(TaskInstance.dag_id == self.dag_id)
         if run_id:
             tis = tis.filter(TaskInstance.run_id == run_id)
         if start_date:
             tis = tis.filter(DagRun.execution_date >= start_date)
 
+        def _task_id_map_filter(val):
+            # Compute a filter  for TI.task_id and TI.map_index based on input values
```
Comment (jedcunningham) [inline airflow/models/dag.py]:
```suggestion
            # Compute a filter for TI.task_id and TI.map_index based on input values
```

**Comment by bbovenzi** [comment]
I won't comment on python code as its not my forté. But, testing with the UI looks correct.

**Comment by ashb** [comment]
@bbovenzi TBH given how fried my brain is this week your python is probably better than mine at the moment!

File: airflow/models/taskinstance.py [hunk 2]
```diff
@@ -228,14 +229,14 @@ def clear_task_instances(
             ti.external_executor_id = None
             session.merge(ti)
 
-        task_id_by_key[ti.dag_id][ti.run_id][ti.try_number].add(ti.task_id)
+        task_id_by_key[ti.dag_id][ti.run_id][ti.map_index][ti.try_number].add(ti.task_id)
```
Comment (uranusjr) [inline airflow/models/taskinstance.py:235]: This mapping is _so_ weird.

File: airflow/models/taskinstance.py [hunk 3]
```diff
@@ -2529,6 +2536,28 @@ def filter_for_tis(tis: Iterable[Union["TaskInstance", TaskInstanceKey]]) -> Opt
...
+        with_map_index = list(filter(lambda v: not isinstance(v, str), vals))
+        filters = []
+
+        if task_id_only:
+            filters.append(cls.task_id.in_(task_id_only))
+        if with_map_index:
+            filters.append(
+                tuple_in_condition((cls.task_id, cls.map_index), with_map_index),
+            )
+        return or_(*filters) if len(filters) > 1 else filters[0]
```
Comment (uranusjr) [inline airflow/models/taskinstance.py]:
```suggestion
        if not filters:
            return true()
        elif len(filters) == 1:
            return filters[0]
        return or_(*filters)
```

Calling `or_()` without any arguments is deprecated https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.or_

Comment (ashb) [reply airflow/models/taskinstance.py]: In my head this function should never result in 0 filters. I should assert that

**Review by jedcunningham — APPROVED** [review]

**Review by uranusjr — APPROVED** [review]
Only part I can’t make sense of is the dag_id-run_id-try_number-map_index lookup. I’ll take your word they are correct.

**Comment by ashb** [comment]
> Only part I can’t make sense of is the dag_id-run_id-try_number-map_index lookup. I’ll take your word they are correct.

Yeah that'd hard to follow, I agree - as the comment hints it was to (try) and build a more efficient query when working on huge dags.

**Comment by uranusjr** [comment]
One failure left

```
tests/api/common/test_mark_tasks.py::TestMarkTasks::test_mark_tasks_subdag
```

(MySQL and MSSQL are green because they skip this test. The logic is likely wrong everywhere.)
